### PR TITLE
fix(behavior_velocity_planner): vehicle points lateral calculation 

### LINKED ISSUE #1734: some packages use only vehicle_width/2
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I've agreed with the maintainers that I can plan this task.

### Description

When calculating the corners of the vehicle, half the width of the vehicle is used as the y point.. This will only work correctly when `left_overhang` and `right_overhang` are equal. Using this formula on non-symmetrical vehicles can lead to many miscalculations.

### Purpose

Correct calculation of the vehicle's points on the y-axis

### Possible approaches

When finding the vehicle's position on the y-axis; `vehicle_width = track_widht + left_overhang + right_overhang`. 
`y = vehicle_width/2` formula is used. but when calculating its position on the y-axis; 
`y = -(right_overhang + (track_width/2))` and `y = left_overhang + (track_width/2)` formulas should be used.

### Definition of done

- [ ] Change y position calculation in ` cv_utils.cpp`
- [ ] Change y position calculation in `run_out/scene.cpp`
- [ ] Change y position calculation in `occlusion_spot/occlusion_spot_utils.cpp`
- [ ] Test

### PR DESCRIPTION
**Pull request by hasanzfdn**
Signed-off-by: Hasanzfdn [hsnzfdn@gmail.com](mailto:hsnzfdn@gmail.com)
## Description
closes #1734 
In addition to the issue, the ``` util.cpp ```code has also been changed

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by takayuki5168** [comment]
LGTM for obstacle avoidance planner

@taikitanaka3 Could you check for behavior velocity planner

**Comment by taikitanaka3** [comment]
Ok I will check this later.

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 1]
```diff
@@ -276,6 +276,9 @@ ObstacleAvoidancePlanner::ObstacleAvoidancePlanner(const rclcpp::NodeOptions & n
     vehicle_param_.wheelbase = vehicle_info.wheel_base_m;
     vehicle_param_.rear_overhang = vehicle_info.rear_overhang_m;
     vehicle_param_.front_overhang = vehicle_info.front_overhang_m;
+    vehicle_param_.right_overhang = vehicle_info.right_overhang_m;
+    vehicle_param_.left_overhang = vehicle_info.left_overhang_m;
+    vehicle_param_.track_width = vehicle_info.wheel_tread_m;
```
Comment (taikitanaka3) [inline planning/obstacle_avoidance_planner/src/node.cpp:281]: Please fix other track with too.
```suggestion
    vehicle_param_.tread_width = vehicle_info.wheel_tread_m;
```

Comment (hasanzfdn) [reply planning/obstacle_avoidance_planner/src/node.cpp:281]: Okay,  I'll fix it now.

File: planning/behavior_velocity_planner/src/scene_module/run_out/scene.cpp [hunk 2]
```diff
@@ -305,12 +307,19 @@ std::vector<geometry_msgs::msg::Point> RunOutModule::createVehiclePolygon(
...
-  const auto p1 = tier4_autoware_utils::calcOffsetPose(base_pose, base_to_front, half_width, 0.0);
-  const auto p2 = tier4_autoware_utils::calcOffsetPose(base_pose, base_to_front, -half_width, 0.0);
-  const auto p3 = tier4_autoware_utils::calcOffsetPose(base_pose, -base_to_rear, -half_width, 0.0);
-  const auto p4 = tier4_autoware_utils::calcOffsetPose(base_pose, -base_to_rear, half_width, 0.0);
+  const double base_to_right =
+    (planner_param_.vehicle_param.wheel_tread / 2.0) + planner_param_.vehicle_param.right_overhang;
+  const double base_to_left =
+    (planner_param_.vehicle_param.wheel_tread / 2.0) + planner_param_.vehicle_param.left_overhang;
+
+  //  const float half_width = planner_param_.vehicle_param.width / 2.0;
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/run_out/scene.cpp:315]: please remove this comment

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 3]
```diff
@@ -246,17 +245,25 @@ ArcCoordinates getOcclusionPoint(const PredictedObject & obj, const ConstLineStr
...
-// calculate value removing offset distance or 0
-double calcSignedLateralDistanceWithOffset(const double lateral, const double offset)
+// calculate lateral distance value to collision point of each object.
+double calcSignedLateralDistanceWithOffset(
+  const double lateral, const double right_overhang, const double left_overhang,
+  const double wheel_tread)
 {
-  // if distance is lower than offset return 0;
-  if (std::abs(lateral) < offset) {
+  //
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:253]: please remove this line too.

File: planning/behavior_velocity_planner/src/scene_module/run_out/scene.cpp (hunk 2 above)
Comment (hasanzfdn) [reply planning/behavior_velocity_planner/src/scene_module/run_out/scene.cpp:315]: I will check unnecessary comments. Thank you

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 4]
```diff
@@ -246,17 +245,25 @@ ArcCoordinates getOcclusionPoint(const PredictedObject & obj, const ConstLineStr
   return arcs.at(0);
 }
 
-// calculate value removing offset distance or 0
-double calcSignedLateralDistanceWithOffset(const double lateral, const double offset)
+// calculate lateral distance value to collision point of each object.
+double calcSignedLateralDistanceWithOffset(
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:249]: @hasanzfdn 
Thank you fot this change but it was a bit hard to read if new change was equal to previous.
Can you refactor this part like this?
```
// calculate lateral distance value to collision point of each object.
double calcSignedLateralDistanceWithOffset(
  const double lateral, const double right_overhang, const double left_overhang,
  const double wheel_tread)
{
  const double offset_left = left_overhang + wheel_tread / 2;
  const double offset_right = right_overhang + wheel_tread / 2;
  if (lateral > 0) {
    return std::max(lateral + offset_left, 0.0);
  }
  // else
  return std::min(lateral - offset_right, 0.0);
}
```

**Comment by hasanzfdn** [comment]
> @hasanzfdn lateral polygon including this PR is something wrong. please check with occlusion spot / run out detecion area polygon image

I tried again and again but I have not encountered such a problem. Screenshot like this.
occlusion_sample_vehicle

**Comment by xmfcx** [comment]
@taikitanaka3 have you tested it again?

**Comment by taikitanaka3** [comment]
@xmfcx 
Sorry I didn't test this again yet,  @xmfcx can you help me testing this PR if you have time to?

**Comment by ismetatabay** [comment]
Thank you @hasanzfdn, can you rebase PR for review? There are some conflicts on obstacle_avoidance_planner.

**Comment by ismetatabay** [comment]
I tested with different maps and vehicles, I think occlusion_spot problem is solved.


Selection_004

Another map and vehicle:

Selection_005

**Review by ismetatabay — APPROVED** [review]
LGTM

**Comment by mitsudome-r** [comment]
@taikitanaka3 @TakaHoribe It looks like LeoDrive member has done a review for this PR. Would you like to run Planning team CI internally before merging?

**Comment by taikitanaka3** [comment]
LGTM for me @TomohitoAndo  can you take a look at run out module?

**Comment by TomohitoAndo** [comment]
@hasanzfdn 
It seems variables of `wheel_tread`, `right_overhang` and `left_overhang` are not assigned.
Could you modify the manager.cpp for run out module like [you did in occlusion spot module](https://github.com/autowarefoundation/autoware.universe/blob/0bca23f52ffc995b4cfac286d35113043f798a41/planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp#L103-L105)?

**Comment by mehmetdogru** [comment]
@hasanzfdn will you be able to work on this or shall we assign this issue to @ismetatabay?

**Comment by ismetatabay** [comment]
I will update the lateral calculation of the vehicle in the run out module.

**Comment by ismetatabay** [comment]
@hasanzfdn Can you update the PR to the draft stage?

**Comment by ismetatabay** [comment]
@TomohitoAndo Hi,  I updated the run out model as you said and it works as in the video you shared.

Selection_008

**Comment by TomohitoAndo** [comment]
@ismetatabay 
Thanks! I checked it works in my environment and CI.
LGTM

**Review by TomohitoAndo — APPROVED** [review]

**Comment by xmfcx** [comment]
@taikitanaka3 a code owner approval is also required, can you check it please?

**Review by xmfcx — APPROVED** [review]

**Pull request merged**